spec/GraphQLOverHTTP.md

```diff
@@ -201,8 +201,38 @@ may not be the GraphQL server but instead some intermediary such as API gateways
 
 ## Status Codes
 
-If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
-If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
-It is recommended to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+If the response has Content-Type GraphQL and contains a non-null `data` entry,
+then it MUST have status code `2xx`, and it SHOULD have status code `200`.
+
+If the response has Content-Type GraphQL and has status code `2xx`,
+the entry `data` entry must be either:
+- equal to `null` (in case of an execution error)
+- not present (in case of a validation error)
+
+The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation typically become part of the result,
+as long as the server is still able to produce a well-formed response.
+
+In case of errors that completely prevent the successful execution of the request,
+the server SHOULD respond with the appropriate status code depending on the concrete error condition.
+
+There are four types of validation errors:
+
+1. Unparseable or invalid request body, for example `NONSENSE` or `{"qeury": "{__typename}"}`
+2. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
+3. Custom document validations that can be performed before execution (e.g. graphql-depth-limit)
+4. Runtime validations performed by resolvers (i.e. during the execution of the GraphQL operation).
+
+The error types 1. and 2. prevent execution of the request entirely and MUST result in
+status code `400` (Bad Request). Types 3. and 4. may still allow partial execution and
+MUST result in status code `200`.
+
+If an unauthenticated client is not allowed to access the schema, the server MUST respond with
+status code `401` (Unauthorized).
+
+If the server failed unexpectedly, it MUST respond with status code `500` (Internal Server Error).
+
+If a client is not authorized to access any of the requested fields, the server MUST respond with
+status code `403` (Forbidden) and not return any `data`.
+If only some requested fields are restricted, and the server can provide at least partial `data`,
+the response MUST have status code `200` (OK) and include `errors` as part of the GraphQL response.
```

Comment:
This should be a non-normative note, as really this is just a description of a partial success. Alternatively remove it.

```suggestion
> NOTE: if certain fields within the GraphQL schema are restricted such that certain clients are not permitted to execute them, access to these fields should result in standard GraphQL field errors following the [error bubbling logic in the GraphQL spec](https://spec.graphql.org/June2018/#sec-Errors-and-Non-Nullability), resulting in a partial-success. As outlined above, this should result in partial `data`, a list of `errors`, and a `2xx` status code, preferably `200` (Okay).
```

Reply:
This was added in contrast to the above statement about returning `403` if all fields are restricted, which I now removed as of your suggestion. Given that the GraphQL spec itself already describes partial success and we mention it in a section above, I will just remove it.